File: searchIndex-platform/module/search-api/search-manager/app/Global.java
import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.Map;

import org.apache.commons.lang3.StringUtils;
import org.ekstep.common.Platform;
import org.ekstep.common.dto.ExecutionContext;
import org.ekstep.common.dto.HeaderParam;
import org.ekstep.common.dto.Response;
import org.ekstep.telemetry.util.TelemetryAccessEventUtil;
import org.ekstep.search.router.SearchRequestRouterPool;
import org.ekstep.telemetry.TelemetryGenerator;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import play.Application;
import play.GlobalSettings;
import play.Logger;
import play.Logger.ALogger;
import play.core.j.JavaResultExtractor;
import play.libs.F.Promise;
import play.mvc.Action;
import play.mvc.Http.Context;
import play.mvc.Http.Request;
import play.mvc.Result;

public class Global extends GlobalSettings {

	private static final ALogger accessLogger = Logger.of("accesslog");
	private static ObjectMapper mapper = new ObjectMapper();

	public void onStart(Application app) {
		System.setProperty("es.set.netty.runtime.available.processors", "false");
		SearchRequestRouterPool.init();
		TelemetryGenerator.setComponent("search-service");
	}

	@SuppressWarnings("rawtypes")
	public Action onRequest(Request request, Method actionMethod) {
		long startTime = System.currentTimeMillis();
		return new Action.Simple() {
			public Promise<Result> call(Context ctx) throws Throwable {
				Promise<Result> call = delegate.call(ctx);
				call.onRedeem((r) -> {
					try {
						String path = request.uri();
						if (!path.contains("/health")) {
							JsonNode requestData = request.body().asJson();
							org.ekstep.common.dto.Request req = mapper.convertValue(requestData,
									org.ekstep.common.dto.Request.class);

							byte[] body = JavaResultExtractor.getBody(r, 0l);
							Response responseObj = mapper.readValue(body, Response.class);

							Map<String, Object> data = new HashMap<String, Object>();
							data.put("StartTime", startTime);
							data.put("Request", req);
							data.put("Response", responseObj);
							data.put("RemoteAddress", request.remoteAddress());
							data.put("ContentLength", body.length);
							data.put("Status", r.status());
							data.put("Protocol", request.secure() ? "HTTPS" : "HTTP");
							data.put("Method", request.method());
							data.put("X-Session-ID", request.getHeader("X-Session-ID"));
							String consumerId = request.getHeader("X-Consumer-ID");
							data.put("X-Consumer-ID", consumerId);
							String deviceId = request.getHeader("X-Device-ID");
							data.put("X-Device-ID", deviceId);
							if(StringUtils.isNotBlank(deviceId))
								ExecutionContext.getCurrent().getGlobalContext().put(HeaderParam.DEVICE_ID.name(),
										deviceId);
							data.put(HeaderParam.APP_ID.name(), request.getHeader("X-App-Id"));

							data.put("X-Authenticated-Userid", request.getHeader("X-Authenticated-Userid"));
							if (StringUtils.isNotBlank(consumerId))
								ExecutionContext.getCurrent().getGlobalContext().put(HeaderParam.CONSUMER_ID.name(),
										consumerId);
							data.put("env", "search");
							data.put("path", path);
							String channelId = request.getHeader("X-Channel-ID");
							if (StringUtils.isNotBlank(channelId))
								ExecutionContext.getCurrent().getGlobalContext().put(HeaderParam.CHANNEL_ID.name(),
										channelId);
							else
								ExecutionContext.getCurrent().getGlobalContext().put(HeaderParam.CHANNEL_ID.name(),
										Platform.config.getString("channel.default"));
							TelemetryAccessEventUtil.writeTelemetryEventLog(data);
							accessLogger.info(request.remoteAddress() + " " + request.host() + " " + request.method()
									+ " " + request.uri() + " " + r.status() + " " + body.length);
						}
					} catch (Exception e) {
						accessLogger.error(e.getMessage());
					}
				});
				return call;
			}
		};
	}
}

File: content-service/content-api/app/Global.scala
import com.typesafe.config.Config
import org.ekstep.content.util.JSONUtils
import org.ekstep.common.dto._
import org.ekstep.learning.router.LearningRequestRouterPool
import org.ekstep.telemetry.TelemetryGenerator
import org.ekstep.telemetry.logger.TelemetryManager
import org.ekstep.telemetry.util.TelemetryAccessEventUtil
import play.api._
import play.api.libs.iteratee.Iteratee
import play.api.mvc._

import scala.concurrent.duration._
import scala.concurrent.{Await, ExecutionContext, Future}
import scala.collection.JavaConverters._


object AccessLoggerFilter extends Filter {

  val accessLogger = Logger("accesslog")

  override def apply(next: RequestHeader => Future[Result])(request: RequestHeader): Future[Result] = {
    implicit val ec: ExecutionContext = ExecutionContext.global

    val startTime = System.currentTimeMillis()
    val resultFuture = next(request)

    resultFuture.foreach(result => {
      val bodyStringFuture = result.body.run(Iteratee.fold(Array.empty[Byte]) { (memo, nextChunk) => memo ++ nextChunk }).map(bytes => new String(bytes))

      val body = Await.result(bodyStringFuture, 1 seconds)

      val responseObj: Response = JSONUtils.deserialize[Response](body.toString)
      val headers = request.headers.toMap
      val data: Map[String, Any] = Map(
        "StartTime"-> startTime,
        "Response"-> responseObj,
      "RemoteAddress"-> request.remoteAddress,
        "ContentLength"-> body.size,
        "path"-> request.uri,
        "Status"-> result.header.status,
        "Protocol"-> (if (request.secure) "HTTPS"
        else "HTTP"),
        "Method"-> request.method,
        "X-Session-ID"-> headers.getOrElse("X-Session-ID", ""),
        "X-Consumer-ID"-> headers.getOrElse("X-Consumer-ID",""),
        "X-Device-ID"-> headers.getOrElse("X-Device-ID",""),
        "X-Authenticated-Userid"-> headers.getOrElse("X-Authenticated-Userid","")
      )

      TelemetryAccessEventUtil.writeTelemetryEventLog(data.asInstanceOf[Map[String, AnyRef]].asJava)
      accessLogger.info(request.remoteAddress + " " + request.host + " " + request.method + " " + request.uri + " " + result.header.status + " " + body.length)

    })

    resultFuture
  }
}


object Global extends WithFilters(AccessLoggerFilter) {

  override def beforeStart(app: Application) {
    val config: Config = play.Play.application.configuration.underlying()
    Logger.info("Application has started...")
  }

  override def onStart(app: Application) = {
    //TODO: check name for component
    TelemetryGenerator.setComponent("content-api")
    TelemetryManager.log("Initialising Request Router Pool")
    LearningRequestRouterPool.init()
    //TODO: Enable local cache updater.
    //TelemetryManager.log("Initialising Local Cache Updater")
    //LocalCacheUpdater.init()
  }


  override def onStop(app: Application) {
    Logger.info("Application shutdown...")
  }

}